OpenSim/Simulation/Model/Smith2016Ligament.cpp

```diff
@@ -0,0 +1,377 @@
+// SCALING
+//=============================================================================
+
+void Smith2016Ligament::extendPostScale(const SimTK::State& s, const ScaleSet& scaleSet)
+{
+    Super::extendPostScale(s, scaleSet);
+
+	GeometryPath& path = upd_GeometryPath();
+	double slack_length = get_slack_length();
+    if (path.getPreScaleLength(s) > 0.0)
+    {
+        double scaleFactor = path.getLength(s) / path.getPreScaleLength(s);
+		set_slack_length(scaleFactor * slack_length);
+
+        // Clear the pre-scale length that was stored in the GeometryPath.
+        path.setPreScaleLength(s, 0.0);
+    }
+}
+
+//=============================================================================
+// GET
+//=============================================================================
+
+double Smith2016Ligament::getTension(const SimTK::State& s) const
+{
+	if (get_appliesForce()) {
+		return getCacheVariableValue<double>(s, "force_total");
+	}
+	else {
+		return 0.0;
+	}
+}
+
+//=============================================================================
+// COMPUTATION
+//=============================================================================
+
+double Smith2016Ligament::computeMomentArm(const SimTK::State& s, Coordinate& aCoord) const
+{
+	return get_GeometryPath().computeMomentArm(s, aCoord);
+}
+
+void Smith2016Ligament::computeForce(const SimTK::State& s,
+							  SimTK::Vector_<SimTK::SpatialVec>& bodyForces,
+							  SimTK::Vector& generalizedForces) const
+{
+	const GeometryPath& path = get_GeometryPath();
+	
+    double slack_length = get_slack_length();
+
+	double length = path.getLength(s);
+	double lengthening_speed = path.getLengtheningSpeed(s);
+
+	double strain = (length - slack_length)/slack_length;
+	double strain_rate = lengthening_speed /slack_length;
+
+	// Calculate Spring Force
+    double force_spring = computeSpringForce(strain);
+
+	// Calculate Damping Force
+	double force_damping = 0.0;
+	if (strain > 0) {
+		force_damping = get_linear_stiffness()*get_normalized_damping_coefficient()*strain_rate;
+	}
+	else {
+		force_damping = 0.0;
+	}
+
+	//Phase-out damping as strain goes to zero with smooth-step function
+	SimTK::Function::Step step(0, 1, 0, 0.01);
+	SimTK::Vector in_vec(1,strain);
+	force_damping = force_damping*step.calcValue(in_vec);
+
+	// total force
+	double force_total = force_spring + force_damping;
+
+	// make sure the ligament is only acting in tension
+	if (force_total < 0.0) {
+		force_total = 0.0;
+	}
+
+	setCacheVariableValue<double>(s,"force_spring", force_spring);
+	setCacheVariableValue<double>(s,"force_damping", force_damping);
+	setCacheVariableValue<double>(s,"force_total", force_total);
+	setCacheVariableValue<double>(s,"length", length);
+	setCacheVariableValue<double>(s,"lengthening_speed", lengthening_speed);
+	setCacheVariableValue<double>(s,"strain", strain);
+	setCacheVariableValue<double>(s,"strain_rate", strain_rate);
```

Comment:
Typically, these cache variables are set lazily: you have a function that gets the value from the cache if the cache is valid, otherwise the value is first calculated and cached.